src/human_encoding/parse/ast.rs

```diff
@@ -0,0 +1,846 @@
+
+use crate::human_encoding::{Error, ErrorSet, Position};
+use crate::jet::Jet;
+use crate::{node, types};
+use crate::{BitIter, Cmr, FailEntropy};
+use santiago::grammar::{Associativity, Grammar};
+use santiago::lexer::{Lexeme, LexerRules};
+
+/// A single non-empty line of a program, of the form x = y :: t
+///
+/// A program is simply a list of such lines
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct Line<J> {
+    /// Position of the first character of the line.
+    pub position: Position,
+    /// The name of the expression being named on the line.
+    pub name: Arc<str>,
+    /// The actual expression, if present (missing for type declarations).
+    pub expression: Option<Expression<J>>,
+    /// The type of the expression, if given (inferred if missing).
+    pub arrow: (Option<Type>, Option<Type>),
+}
+
+/// An expression, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub struct Expression<J> {
+    pub inner: ExprInner<J>,
+    pub position: Position,
+}
+
+impl<J: Jet> Expression<J> {
+    fn reference(name: Arc<str>, position: Position) -> Self {
+        Expression {
+            inner: ExprInner::Reference(name),
+            position,
+        }
+    }
+}
+
+/// An expression, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum ExprInner<J> {
+    /// A reference to another expression
+    Reference(Arc<str>),
+    /// A left assertion (referring to the CMR of an expression on the right)
+    AssertL(Arc<Expression<J>>, AstCmr<J>),
+    /// A right assertion (referring to the CMR of an expression on the left)
+    AssertR(AstCmr<J>, Arc<Expression<J>>),
+    /// An inline expression
+    Inline(node::Inner<Arc<Expression<J>>, J, node::NoDisconnect, node::NoWitness>),
+}
+
+/// A CMR, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum AstCmr<J> {
+    Expr(Arc<Expression<J>>),
+    Literal(Cmr),
+}
+
+/// A type, as represented in the AST
+#[derive(Debug, PartialEq, Eq, Clone, Hash)]
+pub enum Type {
+    /// A named type variable
+    Name(String),
+    /// The unit type 1
+    One,
+    /// The bit type 1+1
+    Two,
+    /// A product type (A * B)
+    Product(Box<Type>, Box<Type>),
+    /// A sum type (A + B)
+    Sum(Box<Type>, Box<Type>),
+    /// An exponential type 2^n where n is one of 2, 4, 8, 16, ..., 1024.
+    TwoN(u32),
+}
+
+impl Type {
+    /// Convert to a Simplicity type
+    pub fn reify(self) -> types::Type {
+        match self {
+            Type::Name(s) => types::Type::free(s),
+            Type::One => types::Type::unit(),
+            Type::Two => types::Type::sum(types::Type::unit(), types::Type::unit()),
+            Type::Product(left, right) => types::Type::product(left.reify(), right.reify()),
+            Type::Sum(left, right) => types::Type::sum(left.reify(), right.reify()),
+            Type::TwoN(n) => types::Type::two_two_n(n as usize), // cast OK as we are only using tiny numbers
```

Comment:
8e711fd6b8a34c454d7ec83a8ad672cc6a8e12f1: Something weird is going on here. `Type::TwoN(n)` means 2^n while `types::Type::two_two_n(n)` means 2^(2^n).

Reply:
`TwoN(n)` is two^(two^n). I should probably rename the variant.